### PR TITLE
Feature/gfs config dirpath fixedfiles

### PR DESCRIPTION
**Pull request by HenryRWinterbottom**
**Description**

This PR addresses issue #1473 .

**Type of change**

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**How Has This Been Tested?**

The directory tree and expected fixed files are defined/copied.

**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

### CODE REVIEW

File: ush/python/pygfs/ufswm/gfs.py [hunk 1]
```diff
@@ -1,20 +1,163 @@
...
+        -----------
+
+        This method configures the base-class object `ufs_config` in
+        accordance with the respective GFS application.
+
+        """
+
+        # HRW: `atmos_run` will need to be defined dynamically once
+        # additional GFS applications are added.
+        config_attr_dict = {'aero_run': 'DO_AERO',
```
Comment (aerorahul) [inline ush/python/pygfs/ufswm/gfs.py:120]: I think we prefer `do_aero` etc.  The less renaming we do, the better.

File: ush/python/pygfs/ufswm/gfs.py [hunk 2]
```diff
@@ -1,20 +1,163 @@
...
+                            'app': 'APP',
+                            'atm_levs': 'LEVS',
+                            'atm_res': "CASE",
+                            'chm_cpl': 'cplchm',
+                            'ice_cpl': 'cplice',
+                            'ice_res': 'ICERES',
+                            'ice_run': 'DO_ICE',
+                            'ocn_cpl': 'cpl',
```
Comment (aerorahul) [inline ush/python/pygfs/ufswm/gfs.py:128]: if `chm_cpl` is `cplchm` for e.g., then why is `ocn_cpl` just `cpl`?

File: ush/python/pygfs/ufswm/gfs.py [hunk 3]
```diff
@@ -1,20 +1,163 @@
...
+        # base-class object `ufs_config`.
+        self.ufs_config.DATA = self._config.DATA
+
+        self.ufs_config.FIX_aer = os.path.join(self._config.FIXgfs, 'aer')
+        self.ufs_config.FIX_am = os.path.join(self._config.FIXgfs, 'am')
+        self.ufs_config.FIX_lut = os.path.join(self._config.FIXgfs, 'lut')
+        self.ufs_config.FIX_orog = os.path.join(self._config.FIXgfs, 'orog')
+        self.ufs_config.FIX_ugwd = os.path.join(self._config.FIXgfs, 'ugwd')
+
+        self.ufs_config.HOMEgfs = self._config.HOMEgfs
```
Comment (aerorahul) [inline ush/python/pygfs/ufswm/gfs.py:105]: Lets think about this:
If we use `gfs` in the name of an object that is "ufs", would that cause issues?
Does `ufs_config` need to know where `HOMEgfs` is?  It needs `FIX` and `parm`, but they could be independent of "HOMEgfs"

File: parm/ufs/fix/gfs/atmos.fixed_files.yaml [hunk 4]
```diff
@@ -1,85 +1,93 @@
-copy:
-  # Atmosphere mosaic file linked as the grid_spec file (atm only)
-  - [$(FIX_orog)/$(atm_res)/$(atm_res)_mosaic.nc, $(DATA)/INPUT/grid_spec.nc]
+dirtree:
```
Comment (aerorahul) [inline parm/ufs/fix/gfs/atmos.fixed_files.yaml:1]: We don't need this `dirtree` section.

File: parm/ufs/fix/gfs/atmos.fixed_files.yaml [hunk 5]
```diff
@@ -1,85 +1,93 @@
...
+          - $(DATA)/RESTART
 
-  # Atmosphere grid tile files
-  - [$(FIX_orog)/$(atm_res)/$(atm_res)_grid.tile1.nc, $(DATA)/INPUT/]
-  - [$(FIX_orog)/$(atm_res)/$(atm_res)_grid.tile2.nc, $(DATA)/INPUT/]
-  - [$(FIX_orog)/$(atm_res)/$(atm_res)_grid.tile3.nc, $(DATA)/INPUT/]
-  - [$(FIX_orog)/$(atm_res)/$(atm_res)_grid.tile4.nc, $(DATA)/INPUT/]
-  - [$(FIX_orog)/$(atm_res)/$(atm_res)_grid.tile5.nc, $(DATA)/INPUT/]
-  - [$(FIX_orog)/$(atm_res)/$(atm_res)_grid.tile6.nc, $(DATA)/INPUT/]
+fixedfiles:
```
Comment (aerorahul) [inline parm/ufs/fix/gfs/atmos.fixed_files.yaml:6]: Keep these as they were before.

File: ush/python/pygfs/ufswm/gfs.py [hunk 6]
```diff
@@ -1,20 +1,163 @@
...
+    TBD
+
+    Parameters
+    ----------
+
+    config: AttrDict
+
+        A Python dictionary containing the run-time configuration
+        attributes.
+
```
Comment (aerorahul) [inline ush/python/pygfs/ufswm/gfs.py:40]: There are no parameters to a class.  They are to the constructor `__init__`.

File: ush/python/pygfs/ufswm/gfs.py [hunk 7]
```diff
@@ -1,20 +1,163 @@
...
+        Description
+        -----------
+
+        This method builds the directory tree and links the relevant
+        files.
+
+        """
+
+        fixedfile_list = [item for item in self.ufs_config.fixedfile_list if (
+            "atmos" in item.lower() or "land" in item.lower())]
```
Comment (aerorahul) [inline ush/python/pygfs/ufswm/gfs.py:72]: This is where a configuration yaml is necessary to elevate this from code to yaml

File: ush/python/pygfs/ufswm/gfs.py [hunk 8]
```diff
@@ -1,20 +1,163 @@
...
         super().__init__("GFS", config)
 
-        # Start putting fixed properties of the GFS
-        self.ntiles = 6
+        self.ufs_config.ntiles = 6
+        self.config_dirtrees()
+        self.config_fcstapp()
+
+    @logit(logger)
+    def build_dirtree(self: UFS) -> None:
```
Comment (aerorahul) [inline ush/python/pygfs/ufswm/gfs.py:61]: Hold this for another PR.

File: ush/python/pygfs/ufswm/ufs.py [hunk 9]
```diff
@@ -50,9 +60,84 @@ def parse_ufs_templates(input_template, output_file, ctx: Dict) -> None:
...
-            print(matches)  # TODO: improve the formatting of this message
-        # TODO: Should we abort here? or continue to write output_file?
+            logger.warn(f"The following variables were not substituted:")
+            logger.warn(f"{', '.join(matches)}.")
 
         with open(output_file, 'w') as fho:
             fho.write(file_out)
+
+    @staticmethod
+    def __get_fixedfile_list(fixedfile_path) -> List:
```
Comment (aerorahul) [inline ush/python/pygfs/ufswm/ufs.py:70]: I don't know why we care about getting a list of all files in a directory.

**Review by aerorahul — CHANGES_REQUESTED** [review]
there is way too much in this PR that
a. we did not discuss in the plan
b. irrelevant to the topic
c. changes that were made that are counter to the ones in the previous PR.

File: parm/ufs/fix/gfs/atmos.fixed_files.yaml (hunk 4 above)
Comment (HenryRWinterbottom) [reply parm/ufs/fix/gfs/atmos.fixed_files.yaml:1]: Does the `FileHandler` know how to build the directory tree prior to copying the respective files?

File: ush/python/pygfs/ufswm/gfs.py (hunk 6 above)
Comment (HenryRWinterbottom) [reply ush/python/pygfs/ufswm/gfs.py:40]: Is this a reference to the documentation?

File: ush/python/pygfs/ufswm/gfs.py (hunk 7 above)
Comment (HenryRWinterbottom) [reply ush/python/pygfs/ufswm/gfs.py:72]: Agreed. I tried to keep this simple for the time-being until we are at that point.

File: ush/python/pygfs/ufswm/gfs.py (hunk 3 above)
Comment (HenryRWinterbottom) [reply ush/python/pygfs/ufswm/gfs.py:105]: Maybe not in the `UFS` class but we could add it in the `GFS` class.

File: ush/python/pygfs/ufswm/gfs.py (hunk 1 above)
Comment (HenryRWinterbottom) [reply ush/python/pygfs/ufswm/gfs.py:120]: That comes from this in the configuration/environment: `DO_AERO False`.

File: parm/ufs/fix/gfs/atmos.fixed_files.yaml (hunk 4 above)
Comment (aerorahul) [reply parm/ufs/fix/gfs/atmos.fixed_files.yaml:1]: No. it does not, but that is not the point of these fix files.
They are containing information about what files are needed, where is the source and what is the destination.  And what is the action (`copy`).
Anything else is not within scope of this file.

File: ush/python/pygfs/ufswm/ufs.py (hunk 9 above)
Comment (HenryRWinterbottom) [reply ush/python/pygfs/ufswm/ufs.py:70]: I thought this was what we discussed. In that path are only fixed-file YAMLs. They are relevant depending on the application of the GFS.

File: parm/ufs/fix/gfs/atmos.fixed_files.yaml (hunk 4 above)
Comment (HenryRWinterbottom) [reply parm/ufs/fix/gfs/atmos.fixed_files.yaml:1]: I am really confused about this. How does the directory tree get built such that the files can be copied? Last time we discussed this piece of the workflow this was needed. Why is it no longer needed?

**Comment by HenryRWinterbottom** [comment]
> there is way too much in this PR that a. we did not discuss in the plan b. irrelevant to the topic c. changes that were made that are counter to the ones in the previous PR.

I am not sure how I am supposed to test the the pieces that were put in place without the remainder of the other bits. My understanding was the following:

- PR #1 -- Add YAML files.
- PR #2 -- Add infrastructure to build directory tree; this includes linking static files.

File: ush/python/pygfs/ufswm/gfs.py (hunk 2 above)
Comment (HenryRWinterbottom) [reply ush/python/pygfs/ufswm/gfs.py:128]: This comes from `DO_COUPLED False` or `cpl False`. Which do you want to use?

**Pull request closed without merging**